//---------------------------------------------------------------------
// Copyright 2020 National Instruments
//
// Licensed under the MIT License;
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://mit-license.org/
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//---------------------------------------------------------------------

//---------------------------------------------------------------------
//---------------------------------------------------------------------
syntax = "proto3";

//---------------------------------------------------------------------
//---------------------------------------------------------------------
option java_multiple_files = true;
option java_package = "labview.measurementservice";
option java_outer_classname = "MeasurementService";
option objc_class_prefix = "LVMS";

//---------------------------------------------------------------------
//---------------------------------------------------------------------
package measurementservice;

//---------------------------------------------------------------------
// The MeasurementService service definition.
//---------------------------------------------------------------------
service MeasurementService {
  
  rpc Invoke (InvokeRequest) returns (InvokeResponse) {}
  
  rpc Query (QueryRequest) returns (QueryResponse) {}
  
  rpc Register (RegistrationRequest) returns (stream ServerEvent) {}
  
  rpc OpenMUXSession (MUXOpenSessionRequest) returns (MUXOpenSessionResponse) {}
  
  rpc CloseMUXSession (MUXCloseSessionRequest) returns (MUXCloseSessionResponse) {}

  rpc OpenMUXChannel (MUXOpenChannelRequest) returns (MUXOpenChannelResponse) {}

  rpc CloseMUXChannel (MUXCloseChannelRequest) returns (MUXCloseChannelResponse) {}
  
  rpc OpenMUXRelay (MUXOpenRelayRequest) returns (MUXOpenRelayResponse) {}
  
  rpc CloseMUXRelay (MUXCloseRelayRequest) returns (MUXCloseRelayResponse) {}
  
  rpc CloseMUXAllRelays (MUXCloseAllRequest) returns (MUXCloseAllResponse) {}
  
  rpc DisconnectAllMUXChannels (MUXDisconnectAllRequest) returns (MUXDisconnectAllResponse) {}
  
  rpc ResetMUX (MUXResetRequest) returns (MUXResetResponse) {}
  
  rpc GetMUXStatus (MUXGetStatusRequest) returns (MUXGetStatusResponse) {}
  
  rpc GetMUXChannelNames (MUXGetChannelNamesRequest) returns (MUXGetChannelNamesResponse) {}
  
  rpc GetMUXRelayInfo (MUXGetRelayInfoRequest) returns (MUXGetRelayInfoResponse) {}
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message ErrorOut {
  int64 errCode = 1;
  string errMessage = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message RelayInfoCluster {
  string RelayName = 1;
  int64 RelayCount = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXOpenSessionRequest {
  string MUXName = 1;
  string MUXTopology = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXOpenSessionResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXCloseSessionRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXCloseSessionResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXOpenChannelRequest {
  string Channel1 = 1;
  string Channel2 = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXOpenChannelResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXCloseChannelRequest {
  string Channel1 = 1;
  string Channel2 = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXCloseChannelResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------

//---------------------------------------------------------------------
message MUXOpenRelayRequest {
  string RelayName = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXOpenRelayResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXCloseRelayRequest {
  string RelayName = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXCloseRelayResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXCloseAllRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXCloseAllResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXDisconnectAllRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXDisconnectAllResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXResetRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXResetResponse {
  ErrorOut error = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXGetStatusRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXGetStatusResponse {
  repeated string Status = 1;
  ErrorOut error = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXGetChannelNamesRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXGetChannelNamesResponse {
  repeated string ChannelNames = 1;
  ErrorOut error = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXGetRelayInfoRequest {
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message MUXGetRelayInfoResponse {
  repeated RelayInfoCluster RelayInfo = 1;
  ErrorOut error = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message InvokeRequest {
  string command = 1;
  string parameter = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message InvokeResponse {
  int64 status = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message QueryRequest {
  string query = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message QueryResponse {
  string message = 1;
  int64 status = 2;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message RegistrationRequest {
  string eventName = 1;
}

//---------------------------------------------------------------------
//---------------------------------------------------------------------
message ServerEvent {
  string eventData = 1;
  int64 serverId = 2;
  int64 status = 3;
}
